include/nuttx/spi/flexspi.h

```diff
@@ -0,0 +1,521 @@
+                                                      */
+
+  FLEXSPI_AHB_WRITE_WAIT_UNIT128_AHB_CYCLE   = 0x3,  /* AWRWAIT unit is 128
+                                                      * ahb clock cycle.
+                                                      */
+
+  FLEXSPI_AHB_WRITE_WAIT_UNIT512_AHB_CYCLE   = 0x4,  /* AWRWAIT unit is 512
+                                                      * ahb clock cycle.
+                                                      */
+
+  FLEXSPI_AHB_WRITE_WAIT_UNIT2048_AHB_CYCLE  = 0x5,  /* AWRWAIT unit is 2048
+                                                      * ahb clock cycle.
+                                                      */
+
+  FLEXSPI_AHB_WRITE_WAIT_UNIT8192_AHB_CYCLE  = 0x6,  /* AWRWAIT unit is 8192
+                                                      * ahb clock cycle.
+                                                      */
+
+  FLEXSPI_AHB_WRITE_WAIT_UNIT32768_AHB_CYCLE = 0x7,  /* AWRWAIT unit is 32768
+                                                      * ahb clock cycle.
+                                                      */
+};
+
+/* External device configuration items */
+
+struct flexspi_device_config_s
+{
+  uint32_t flexspi_root_clk; /* FLEXSPI serial root clock */
+  bool is_sck2_enabled;      /* FLEXSPI use SCK2 */
+  uint32_t flash_size;       /* Flash size in KByte */
+
+  enum flexspi_cs_interval_cycle_unit_e cs_interval_unit; /* CS interval unit, 1
+                                                      * or 256 cycle.
+                                                      */
+
+  uint16_t cs_interval;     /* CS line assert interval, multiply CS
+                             * interval unit to get the CS line assert
+                             * interval cycles.
+                             */
+
+  uint8_t cs_hold_time;     /* CS line hold time */
+  uint8_t cs_setup_time;    /* CS line setup time */
+  uint8_t data_valid_time;  /* Data valid time for external device */
+  uint8_t columnspace;      /* Column space size */
+  bool enable_word_address; /* If enable word address */
+  uint8_t awr_seq_index;    /* Sequence ID for AHB write command */
+  uint8_t awr_seq_number;   /* Sequence number for AHB write command */
+  uint8_t ard_seq_index;    /* Sequence ID for AHB read command */
+  uint8_t ard_seq_number;   /* Sequence number for AHB read command */
+
+  enum flexspi_ahb_write_wait_unit_e ahb_write_wait_unit;  /* AHB write wait unit */
+
+  uint16_t ahb_write_wait_interval;  /* AHB write wait interval, multiply AHB
+                                      * write interval unit to get the AHB
+                                      * write wait cycles.
+                                      */
+
+  bool enable_write_mask;   /* Enable/Disable FLEXSPI drive DQS pin as write mask
+                             * when writing to external device.
+                             */
+};
+
+/* The FlexSPI vtable */
+
+struct flexspi_dev_s;
+struct flexspi_ops_s
```

Comment:
One question: is flexspi specific to NXP or exist on other SoC?

Reply:
FlexSPI is currently NXP specific, although many different families of CPU support it. It is like QSPI with additional capabilities.

Reply:
@jturnsek  Given that most all the the NXP IP block carrier the "flex" prefix. Can we determine the actual differences between QSPI and flexSPI abstract the interfaces to be vendor neutral in definition and naming?

Reply:
@davids5 I was thinking how to use qspi structures to implement flexspi driver (and it is possible, but not efficient). I decided then to take an implementation from the Zephyr and use it as a base for this driver. As you can see from the nuttx/sp/flexspi.h header it has a different way of setting the peripheral, thus I think FlexSPI should be a separate structure from QSPI. Current implementation is also very basic and it can grow through time to support HyperBus interface or for example multiple physical NOR devices on the same FlexSPI peripheral. About being vendor neutral in definition and naming, I don't know. I think FlexSPI is quite a good name for it, although it is NXP related.

Reply:
The file is arch (ahb references) and vendor specific.  Possibly @patacongo can give you some insight on a approach to use here.

Reply:
@davids5 These are all valid points. I agree the flexspi.h device structure holds some vendor specific naming in order to setup the bus as needed for various flash devices or FPGAs. This being said, it can control many different devices and this is the strong point here. The driver could be implemented much more vendor agnostic, but this would bring in a transformation layer for all those bus settings and it would take a considerable amount of time to first think about the interface and then implement it. My opinion is that this is uneccessary for this vendor specific peripheral,because there are quite a lot of NXP boards with the i.MX RT CPU in the NX which could benefit from this driver. 

The other option here is to go with the general QSPI driver API and to sacrify the efficiency of the transfers.

Reply:
@jturnsek believe me I hear you. My concern it the coming from preserving the NuttX [INVIOLABLES]( https://github.com/apache/incubator-nuttx/blob/master/INVIOLABLES.md) hence , I have asked @patacongo to add guidance here.

Reply:
I think we should integrate this driver because it will be useful for some users, even although it is an arch specific driver. It is better to have FlexSPI support hardware specifically for iMXRT than doesn't have any support. What do you think @xiaoxiang781216 and @davids5 ?

Reply:
@acassis I agree it is better to have it. The issues are: Is it in the correct place in the source tree? and does is violate any of the  INVIOLABLES. Do you have an opinion on these 2 issues?

Reply:
@davids5  You are right! I was focusing in the flexspi driver itself! @jturnsek Could you please move the files from driver/mtd/ to arch/arm/src/imxrt ? See the arch/xtensa/src/esp32/esp32_spiflash.c as example, it is an arch specific driver to support ESP32 SPIFLASH. It uses the MTD but doesn't include itself inside driver/mtd/

Reply:
The problem is that the driver number(flash variant x spi variant) will increase to unmanage the situation with the current approach. Can we design the spi driver interface into two part:

1. The common callbacks(e.g. transfer, command...) which is directly called by flash driver have the same prototype
2. The special callbacks(e.g. config, ...) could have the different prototype(spi, qspi, flexspi), and call these function from board file.

So, we can keep the flash and spi combination in the linear complexity(n+m), instead the square complexity(n x m).

Reply:
@xiaoxiang781216 I agree with you. When I was looking how to implement the driver, I have noticed that there is already two interfaces (SPI and QSPI), which in my opinion are more similar then QSPI and FlexSPI, thus I was thinking there won't be a problem to introduce (as the Nuttx beginer) the third one. I apologize for not respecting the INVIOLABLES.

@acassis Alan you are right. I could move everything into board and arch, thus it would be completely separated from the mtd layer.

Reply:
@jturnsek it isn't your problem. Since you are working on FlexSPI interface now, it's time to consider whether there is a better approach to generalize SPI interface around SPI/QSPI/FlexSPI at least for transfer, so other subsystem(flash, sensor) could talk with one common spi driver interface.

Reply:
I have followed Alan's advice and move the FlexSPI driver from the mtd to the MIMXRT1064-EVK board directory. This way it is completely isolated from the rest of the code.See the #3462. Although FlexSPI could be used for many peripherals, it is used mainly for NOR flash devices present on the i.MX RT boards.

Reply:
Ok, the change at least don't make the public interface worse.

Reply:
It is fine now! LGTM!